include/nbl/asset/IRayTracingPipeline.h

```diff
@@ -0,0 +1,173 @@
+        using SpecInfo = ShaderType::SSpecInfo;
+        template<typename ExtraLambda>
+        inline bool impl_valid(ExtraLambda&& extra) const
+        {
+          if (!IPipeline<PipelineLayoutType>::SCreationParams::layout)
+            return false;
+
+          core::bitflag<ICPUShader::E_SHADER_STAGE> stagePresence = {};
+          for (const auto info : shaders)
+            if (info.shader)
+            {
+              if (!extra(info))
+                return false;
+              const auto stage = info.shader->getStage();
+              if ((stage & ~ICPUShader::E_SHADER_STAGE::ESS_ALL_RAY_TRACING)!=0)  
+                return false;
+              if (!std::has_single_bit<std::underlying_type_t<ICPUShader::E_SHADER_STAGE>>(stage))
+                return false;
+              stagePresence |= stage;
+            }
+
+          auto getShaderStage = [this](size_t index) -> ICPUShader::E_SHADER_STAGE
+            {
+              return shaders[index].shader->getStage();
+            };
+
+          if (shaderGroups.raygen.index >= shaders.size())
+            return false;
+          if (getShaderStage(shaderGroups.raygen.index) != ICPUShader::E_SHADER_STAGE::ESS_RAYGEN)
+            return false;
+
+          auto isValidShaderIndex = [this, getShaderStage](size_t index, ICPUShader::E_SHADER_STAGE expectedStage) -> bool
+            {
+              if (index == SShaderGroupsParams::SIndex::Unused)
+                return true;
+              if (index >= shaders.size())
+                return false;
+              if (getShaderStage(index) != expectedStage)
+                return false;
+              return true;
+            };
+
+          for (const auto& shaderGroup : shaderGroups.hits)
+          {
+            if (!isValidShaderIndex(shaderGroup.anyHit, ICPUShader::E_SHADER_STAGE::ESS_ANY_HIT))
+              return false;
+
+            if (!isValidShaderIndex(shaderGroup.closestHit, ICPUShader::E_SHADER_STAGE::ESS_CLOSEST_HIT))
+              return false;
+
+            if (!isValidShaderIndex(shaderGroup.intersectionShader, ICPUShader::E_SHADER_STAGE::ESS_INTERSECTION))
+              return false;
+          }
+
+          for (const auto& shaderGroup : shaderGroups.misses)
+          {
+            if (!isValidShaderIndex(shaderGroup.index, ICPUShader::E_SHADER_STAGE::ESS_MISS))
+              return false;
+          }
+
+          for (const auto& shaderGroup : shaderGroups.callables)
+          {
+            if (!isValidShaderIndex(shaderGroup.index, ICPUShader::E_SHADER_STAGE::ESS_CALLABLE))
+              return false;
+          }
+          return true;
+        }
+
+      public:
+        inline bool valid() const
+        {
+          return impl_valid([](const SpecInfo& info)->bool
+          {
+            if (!info.valid())
+              return false;
+            return false;
+          });
+        }
+
+        std::span<const SpecInfo> shaders = {};
+        SShaderGroupsParams shaderGroups;
```

Comment:
I'm not going to ask you to rework this now, but when we get onto #839 we could just not specify the `shaders` and just have `shaderGroups` without indices referencing `IShader` + specialization info directly

Then we'd use a `unordered_map` or `map` to implicitly produce the `shaders` argument during pipeline creation.

Reply:
However right now only the flags should stay in `IRayTracingPipeline`, the CreationParams (especially the validation) should go into `IGPURayTracingPipeline`

because the `ICPURayTracingPipeline` we should probably do differently (shaders in vectors) and the validation shouldn't be as aggressive

It should be possible to add/remove shader groups to `ICPURayTracingPipeline` after creation and make it "somewhat" incompatible with the NO_NULL flag (the asset converter can "patch" them out)